src/trans/gpu/internal/tpm_fftc.F90

```diff
@@ -19,170 +20,159 @@ MODULE TPM_FFTC
+  REAL(KIND=C_FLOAT), INTENT(OUT) :: PREEL_COMPLEX(:)
+  INTEGER(KIND=JPIM),INTENT(IN)  :: KFIELD
+  INTEGER(KIND=JPIM),INTENT(IN)  :: LOENS(:), OFFSETS(:)
+  INTERFACE
+    SUBROUTINE EXECUTE_DIR_FFT_FLOAT_C(PREEL_REAL,PREEL_COMPLEX,KFIELD,LOENS,OFFSETS,NFFT) BIND(C, name="execute_dir_fft_float")
+      USE ISO_C_BINDING
+      REAL(KIND=C_FLOAT), INTENT(IN) :: PREEL_REAL(:)
+      REAL(KIND=C_FLOAT), INTENT(OUT) :: PREEL_COMPLEX(:)
+      INTEGER(KIND=C_INT),INTENT(IN),VALUE  :: KFIELD
+      INTEGER(KIND=C_INT),INTENT(IN)  :: LOENS(:), OFFSETS(:)
+      INTEGER(KIND=C_INT),INTENT(IN),VALUE :: NFFT
+    END SUBROUTINE
+  END INTERFACE
+
+  !$ACC HOST_DATA USE_DEVICE(PREEL_REAL,PREEL_COMPLEX)
+  CALL EXECUTE_DIR_FFT_FLOAT_C(PREEL_REAL,PREEL_COMPLEX,KFIELD,LOENS,OFFSETS,SIZE(LOENS))
+  !$ACC END HOST_DATA
+
+END SUBROUTINE
+SUBROUTINE EXECUTE_DIR_FFT_DOUBLE(PREEL_REAL,PREEL_COMPLEX,KFIELD,LOENS,OFFSETS)
+  USE PARKIND_ECTRANS ,ONLY : JPIM
+
+  IMPLICIT NONE
+
+  REAL(KIND=C_DOUBLE), INTENT(IN) :: PREEL_REAL(:)
+  REAL(KIND=C_DOUBLE), INTENT(OUT) :: PREEL_COMPLEX(:)
+  INTEGER(KIND=JPIM),INTENT(IN)  :: KFIELD
+  INTEGER(KIND=JPIM),INTENT(IN)  :: LOENS(:), OFFSETS(:)
+  INTERFACE
+    SUBROUTINE EXECUTE_DIR_FFT_DOUBLE_C(PREEL_REAL,PREEL_COMPLEX,KFIELD,LOENS,OFFSETS,NFFT) BIND(C, name="execute_dir_fft_double")
+      USE ISO_C_BINDING
+      REAL(KIND=C_DOUBLE), INTENT(IN) :: PREEL_REAL(:)
+      REAL(KIND=C_DOUBLE), INTENT(OUT) :: PREEL_COMPLEX(:)
+      INTEGER(KIND=C_INT),INTENT(IN),VALUE  :: KFIELD
+      INTEGER(KIND=C_INT),INTENT(IN)  :: LOENS(:), OFFSETS(:)
+      INTEGER(KIND=C_INT),INTENT(IN),VALUE :: NFFT
+    END SUBROUTINE
+  END INTERFACE
+
+  !$ACC HOST_DATA USE_DEVICE(PREEL_REAL,PREEL_COMPLEX)
+  CALL EXECUTE_DIR_FFT_DOUBLE_C(PREEL_REAL,PREEL_COMPLEX,KFIELD,LOENS,OFFSETS,SIZE(LOENS))
+  !$ACC END HOST_DATA
+
+END SUBROUTINE
+
+SUBROUTINE EXECUTE_INV_FFT_FLOAT(PREEL_COMPLEX,PREEL_REAL,KFIELD,LOENS,OFFSETS)
+  USE PARKIND_ECTRANS ,ONLY : JPIM
+
+  IMPLICIT NONE
+
+  REAL(KIND=C_FLOAT), INTENT(IN) :: PREEL_COMPLEX(:)
+  REAL(KIND=C_FLOAT), INTENT(OUT) :: PREEL_REAL(:)
+  INTEGER(KIND=JPIM),INTENT(IN)  :: KFIELD
+  INTEGER(KIND=JPIM),INTENT(IN)  :: LOENS(:), OFFSETS(:)
+  INTERFACE
+    SUBROUTINE EXECUTE_INV_FFT_FLOAT_C(PREEL_COMPLEX,PREEL_REAL,KFIELD,LOENS,OFFSETS,NFFT) BIND(C, name="execute_inv_fft_float")
+      USE ISO_C_BINDING
+      REAL(KIND=C_FLOAT), INTENT(IN) :: PREEL_COMPLEX(:)
+      REAL(KIND=C_FLOAT), INTENT(OUT) :: PREEL_REAL(:)
+      INTEGER(KIND=C_INT),INTENT(IN),VALUE  :: KFIELD
+      INTEGER(KIND=C_INT),INTENT(IN)  :: LOENS(:), OFFSETS(:)
+      INTEGER(KIND=C_INT),INTENT(IN),VALUE :: NFFT
+    END SUBROUTINE
+  END INTERFACE
+
+  !$ACC HOST_DATA USE_DEVICE(PREEL_COMPLEX,PREEL_REAL)
+  CALL EXECUTE_INV_FFT_FLOAT_C(PREEL_COMPLEX,PREEL_REAL,KFIELD,LOENS,OFFSETS,SIZE(LOENS))
+  !$ACC END HOST_DATA
+END SUBROUTINE
+
+SUBROUTINE EXECUTE_INV_FFT_DOUBLE(PREEL_COMPLEX,PREEL_REAL,KFIELD,LOENS,OFFSETS)
```

Comment:
The FFT plan mangagement was moved to the CUDA code (because it not only a FFT plan cache anymore, but actually a CUDA graph+fft plan cache)